text/0000-from-lossy.md

```diff
@@ -0,0 +1,321 @@
+- lossy conversions (e.g. `i64` → `f32`)
+- lossy fallible conversions (e.g. `f64` → `u64`)
+- truncations on unsigned integers (e.g. `u64` → `u32`, dropping unused high bits)
+- sign-ignoring coercions/transmutations (e.g. `i8` → `u8`, `i64` → `i32`);
+  these can yield totally different values due to interpretation of the sign
+  bit (e.g. `3i32 << 14` is 49152=2^14+2^15; converting to `i16` yields -16384=2^14-2^15)
+- conversions between types with platform-dependent size (i.e. `usize` and `isize`)
+
+This RFC is principally concerned with lossy conversions, but considers other
+conversions for a broader picture.
+
+It is *my opinion* that we should cover all *common* conversions performed by
+the `as` keyword redundant with `std` library functionality, and *consider*
+deprecating `as` eventually. Whether this entails traits covering all the above
+conversions or methods on primitive types for some conversions, or even leaves
+no alternative to `transmute` in some cases is beyond the scope of this RFC.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+Type conversions can be handled by the following traits:
+
+- `From` for infallible, exact conversions (e.g. widening, `u8` → `u16`)
+- `TryFrom` for fallible, exact conversions (e.g. narrowing, `u16` → `u8`, and signed, `i16` → `u16`)
+- `FromLossy` for infallible, lossy conversions (mostly concerning floating-point types, e.g. `u32` → `f32`)
+- `TryFromLossy` for fallible, inexact conversions (e.g. `f32` → `u32`)
+- `TruncateFrom` for truncations (e.g. `u16` → `u8` which drops high bits)
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+### `From` trait
+
+Tweak the documentation to clarify this.
+
+Remove several implementations on SIMD types (currently including many fallible
+conversions like `f32x4` → `i8x4`; nightly only).
+
+### `FromLossy` trait
+
+Add `std::convert::FromLossy`:
+
+```rust
+/// A trait for conversions which are potentially inexact.
+/// 
+/// It is required that conversions do not fail, and required that results are
+/// either approximately equivalent or are an appropriate special value.
+/// For example, it is reasonable for an implementation converting `f64` to
+/// `f32` to lose precision and to return positive or negative infinity for
+/// out-of-range results, as well as to drop payloads from NaN values.
+pub trait FromLossy {
+    fn from_lossy(x: T) -> Self;
+}
+```
+
+Add implementations to `f64` from all of:
+
+- `u64, i64, u128, i128`
+
+and to `f32` from all of:
+
+- `f64, u32, i32, u64, i64, u128, i128`
+
+(Note: other integer → float conversions are
+already handled by `From`. There is a question below about trait overlap.)
+
+Where conversions are not exact, they should be equivalent to a conversion
+which first forces some number of low bits of the integer representation to
+zero, and then does an exact conversion to floating-point.
```

Comment:
Is this trying to avoid specifying a rounding mode? Or is it trying to specify a rounding mode in some convoluted manner? Either way, all these conversions should be specified to round *to nearest, ties to even*. That's **the** default rounding mode and we (as well as other languages) use it everywhere except for float -> int conversions.

Reply:
True, this is convoluted. And you're correct, since the output type here is floating-point, it would be expected to round to nearest with ties to even.